### ISSUE TITLE
Selector inconsistency between MRI and JRuby

### ISSUE THREAD

**Issue by kirs** [sparklemotion/nokogiri#1653]
Nokogiri version: 1.8.0

Script:

```ruby
require 'bundler/setup'
require 'nokogiri'

str = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><root><area><p>content</p></area></root>"
doc = Nokogiri::HTML::Document.parse(str)
result = doc.root.css("root > area + p")
puts result.inspect
```

On MRI 2.3.3:

```
$ ruby nok.rb
[#<Nokogiri::XML::Element:0x3ff7d25d7730 name="p" children=[#<Nokogiri::XML::Text:0x3ff7d25d7528 "content">]>]
```

On jruby-9.1.10.0:

```
$ jruby --dev nok.rb
[]
```

Expected: the value to be consistent across MRI and JRuby.

**Comment by flavorjones** [comment]
Apologies for the embarrassingly long response time on this.

TL;DR, what you're seeing is actually __parsing__ differences, and not __query selector__ differences. It may be useful to point out that Nokogiri uses different parsers on JRuby and MRI.

I added `puts doc` to your script, to emit the parsed document structure: On JRuby, NekoHTML parses this as:

```
# jruby / nekohtml
<html>
  <head></head>
  <body>
    <root><map><area><p>content</p></map></root>
  </body>
</html>
```

while on MRI, libxml2 parses this as:

```
# mri / libxml2
<html>
  <body>
    <root><area><p>content</p></root>
  </body>
</html>
```

This is simply because every HTML parser "fixes" "broken" markup in slightly different ways. In this case, JRuby is correctly adding a `<map>` ancestor for the `<area>` tag, while libxml2 just leaves the document structure as-is.

Info on the `area` tag is here:

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/area

Notably:

>  Permitted parents | Any element that accepts phrasing content. The <area>element must have an ancestor <map>, but it need not be a direct parent.

I hope this makes sense? Apologies again for the long reply time.

**Comment by kirs** [comment]
This makes sense, thanks for the background!
